3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala

```diff
@@ -0,0 +1,206 @@
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.enrich.common.utils
+
+import cats.Monad
+import cats.data.{EitherT, NonEmptyList, Validated}
+import cats.effect.Clock
+import cats.implicits._
+
+import io.circe.Json
+
+import com.snowplowanalytics.iglu.client.{Client, ClientError}
+import com.snowplowanalytics.iglu.client.resolver.registries.RegistryLookup
+
+import com.snowplowanalytics.iglu.core.circe.instances._
+import com.snowplowanalytics.iglu.core.{ParseError, SchemaCriterion, SchemaKey, SelfDescribingData}
+
+import com.snowplowanalytics.snowplow.badrows.FailureDetails
+
+import com.snowplowanalytics.snowplow.enrich.common.outputs.EnrichedEvent
+
+/** Helper to validate:
+ *  - An unstructured event,
+ *  - The input contexts of an event,
+ *  - The contexts adveuxded by the enrichments.
+ */
+object IgluUtils {
+
+  /** Extract unstructured event from event and validate against its schema
+   *  @param event Snowplow enriched event to extract unstructured event from (in String)
+   *  @param client Iglu client used for SDJ validation
+   *  @param field Name of the field containing the unstructured event, to put in the bad row
+   *               in case of failure
+   *  @param criterion Expected schema for the JSON containing the unstructured event
+   *  @return Valid unstructured event if the input event has one or [[None]]
+   */
+  def extractAndValidateUnstructEvent[F[_]: Monad: RegistryLookup: Clock](
+    event: EnrichedEvent,
+    client: Client[F, Json],
+    field: String = "ue_properties",
+    criterion: SchemaCriterion = SchemaCriterion("com.snowplowanalytics.snowplow", "unstruct_event",
+      "jsonschema", 1, 0)
+  ): F[Validated[FailureDetails.SchemaViolation, Option[SelfDescribingData[Json]]]] =
+    (Option(event.unstruct_event) match {
+      case Some(rawUnstructEvent) =>
+        for {
+          // Validate input Json string and extract unstructured event
+          unstruct <- extractInputData(rawUnstructEvent, field, criterion, client)
+          // Parse Json unstructured event as SelfDescribingData[Json]
+          unstructSDJ <- SelfDescribingData
+            .parse(unstruct)
+            .leftMap(e => FailureDetails.SchemaViolation.NotIglu(unstruct, e))
+            .toEitherT[F]
+          // Check SelfDescribingData[Json] of unstructured event
+          _ <- check(client, unstructSDJ)
+            .leftMap(e => FailureDetails.SchemaViolation.IgluError(e._1, e._2))
+            .leftWiden[FailureDetails.SchemaViolation]
```

Comment:
Could someone of @snowplow/com-snowplowanalytics-engineering-datacapability explain to me why I need `leftWiden` here but not on line 63 please?